text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+
+The procedure for handling the equivocation catastrophes will have to consist of two main parts: what the software is expected to do in such a case, and what are the manual actions required to proceed.
+
+The obvious first thing that should happen in the case of an equivocation catastrophe is that the software should alert the users about the situation. The node software is aware of the configured FTT, so when it detects equivocations, it can notice if the total weight of the equivocators exceeds the current FTT. When it is exceeded, the users should be made aware of it.
+
+In order to make sure that as many validators as possible become aware of all the equivocations eventually, the operation of the nodes shouldn't stop. Finalizing blocks should be halted, as finalization ceases to mean anything when it can be reverted, but the nodes should continue gossiping votes.
+
+The nodes could also output the hash of the last block finalized before FTT has been exceeded and all the transactions included in the later blocks, and the list of equivocators. These will be useful for the manual steps.
+
+The manual actions afterwards will consist of two main steps:
+
+1. Reaching social consensus about the new starting point,
+2. Restarting the network with the new starting point.
+
+We will go into more detail about these steps below.
+
+### Social consensus
+
+For the initial version of this CEP, we assume that we have some kind of platform that enables open communication between validators and CasperLabs. The chosen platform should allow for storing, linking, referencing and displaying the whole public process and communication. We also should make sure that the content is in a format that can be displayed elsewhere.
+
+The first essential step will be determining the starting point for the network after the restart. The suggested approach here would be to agree upon the new initial global state, new set of validators and restart the network as if it was starting from scratch.
+
+A desirable goal for the initial global state would be to revert as few transactions from all the forks as possible, in order to limit the damage. Every transaction that was included in a block which appeared to be finalized for a time could have caused real world consequences (like, for example, a merchant sending ordered goods). Reverting any such transaction might potentially cause a loss to a user.
+
+In order to minimize losses, the following approach is proposed: first, choose the last finalized block as the starting point. If the equivocators caused multiple forks to appear, choose the tip of one of the forks. The exact criteria of choice will be left to the community to decide.
+
+After the block is chosen, take its global state and all transactions from the blocks in other forks and attempt to replay them on top of that global state. Since the order could matter, the community would have to agree on one. Call the resulting global state the "pre-initial state".
```

Comment:
This is something we need to provide the tools for, so that — without having to develop a whole new piece of software — the community can choose, define and configure that pre-initial state.
E.g. it could be defined by:
* a known block from before the equivocation,
* a list of blocks whose deploys should be replayed on top of that, and
* a list of system transactions to be run on top of that — e.g. to slash and/or refund (in case of bugs).

(But I guess the third point could be done in a separate step after the restart, as an upgrade point.)